netspeed/src/backend.c

```diff
@@ -327,3 +350,413 @@ get_wireless_info (DevInfo *devinfo)
+        char buf[100];
+        char *pos = buf;
+
+        pos += sprintf(pos, _("%d (%d MHz)"),
+                       ieee80211_frequency_to_channel (freq), freq);
+
+        if (tb_msg[NL80211_ATTR_CHANNEL_WIDTH])
+            pos += sprintf(pos, _(", width: %s"),
+                           channel_width_name (nla_get_u32 (tb_msg[NL80211_ATTR_CHANNEL_WIDTH])));
+
+        pos = '\0';
+        devinfo->channel = g_strdup (buf);
+    }
+
+    return NL_SKIP;
+}
+
+void
+get_wireless_info (DevInfo *devinfo)
+{
+    struct nl80211_state nlstate;
+    int err, ret;
+    struct nl_msg *msg;
+    signed long long devidx = 0;
+
+    err = nl80211_init (&nlstate);
+    if (err) {
+        g_warning ("ToDo");
+        return;
+    }
+
+    devidx = if_nametoindex (devinfo->name);
+
+    /* Get MAC, SSID & channel info from interface message */
+    msg = nlmsg_alloc ();
+    if (!msg) {
+        g_warning ("failed to allocate netlink message");
+        goto cleanup;
+    }
+    genlmsg_put (msg, 0, 0, nlstate.nl80211_id, 0, NLM_F_DUMP, NL80211_CMD_GET_INTERFACE, 0);
+    /* Add message attribute, which interface to use */
+    nla_put_u32 (msg, NL80211_ATTR_IFINDEX, devidx);
+    /* Add the callback */
+    nl_socket_modify_cb (nlstate.nl_sock, NL_CB_VALID, NL_CB_CUSTOM, iface_cb, devinfo);
+    /* Send the message */
+    ret = nl_send_auto (nlstate.nl_sock, msg);
+    g_debug ("NL80211_CMD_GET_INTERFACE sent %d bytes to the kernel", ret);
+    /* Retrieve the kernel's answer */
+    ret = nl_recvmsgs_default (nlstate.nl_sock);
+    nlmsg_free (msg);
+    if (ret < 0) {
+        g_warning ("failed to recive netlink message");
+    }
+
+    if (!devinfo->running)
+        goto cleanup;
+
+    /* Get station MAC from scan message */
+    msg = nlmsg_alloc ();
+    if (!msg) {
+        g_warning ("failed to allocate netlink message");
+        goto cleanup;
+    }
+    genlmsg_put (msg, 0, 0, nlstate.nl80211_id, 0, NLM_F_DUMP, NL80211_CMD_GET_SCAN, 0);
+    /* Add message attribute, which interface to use */
+    nla_put_u32 (msg, NL80211_ATTR_IFINDEX, devidx);
+    /* Add the callback */
+    nl_socket_modify_cb (nlstate.nl_sock, NL_CB_VALID, NL_CB_CUSTOM, scan_cb, devinfo);
+    /* Send the message */
+    ret = nl_send_auto (nlstate.nl_sock, msg);
+    g_debug ("NL80211_CMD_GET_SCAN sent %d bytes to the kernel", ret);
+    /* Retrieve the kernel's answer */
+    ret = nl_recvmsgs_default (nlstate.nl_sock);
+    nlmsg_free (msg);
+    if (ret < 0) {
+        g_warning ("failed to recive netlink message");
+        goto cleanup;
+    }
+
+    if (!devinfo->running)
+        goto cleanup;
+
+    /* Get in/out bitrate/rate/total, siganl quality from station message */
+    msg = nlmsg_alloc ();
+    if (!msg) {
+        g_warning ("failed to allocate netlink message");
+        goto cleanup;
+    }
+    genlmsg_put (msg, 0, 0, nlstate.nl80211_id, 0, NLM_F_DUMP, NL80211_CMD_GET_STATION, 0);
```

Comment:
There is a build warning. https://travis-ci.org/github/mate-desktop/mate-applets/jobs/705601948#L4769
```
backend.c: In function 'get_wireless_info':

backend.c:743:5: warning: '*((void *)&nlstate+8)' may be used uninitialized in this function [-Wmaybe-uninitialized]

  743 |     genlmsg_put (msg, 0, 0, nlstate.nl80211_id, 0, NLM_F_DUMP, NL80211_CMD_GET_STATION, 0);

      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
done